src/ucs/sys/topo.h

```diff
@@ -0,0 +1,100 @@
+/**
+* Copyright (C) NVIDIA Corporation. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_TOPO_H
+#define UCS_TOPO_H
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <ucs/type/status.h>
+#include <limits.h>
+
+BEGIN_C_DECLS
+
+/** @file topo.h */
+
+/**
+ * @ingroup UCS_RESOURCE
+ * System Device abstraction
+ */
+typedef struct ucs_sys_device {
+    unsigned int id;             /**< Index of the unit */
+    unsigned int bus_id;         /**< bus ID of of the device if applicable*/
+    unsigned int numa_node;      /**< NUMA node assoicated with the device*/
+    char         name[PATH_MAX]; /**< Name of the device*/
+} ucs_sys_device_t;
+
+
+/**
+ * PCIe distance categories and answers what needs to be crossed to reach
+ */
+typedef enum {
+    UCS_DISTANCE_PIX = 0, /* traverse 1 PCIe switch */
+    UCS_DISTANCE_PXB,     /* traverse >1 PCIe switches */
+    UCS_DISTANCE_PHB,     /* traverse host bridge */
+    UCS_DISTANCE_NODE,    /* traverse host bridge in the same numa node */
+    UCS_DISTANCE_SYS      /* traverse CPU interconnect (like QPI) */
+} ucs_distance_t;
```

Comment:
The problem is that the actual latency and bandwidth is very architecture specific (for example, whether we're using Skylake vs Broadwell). This essentially translates to maintaining architecture specific information for the selection process. An alternative is to ignore the reported latency/bandwidth (should the architecture not have an entry in the internal repo) and just use information about relative distance in the PCIe subsystem and make decisions even at the UCP-level.

Reply:
then, topology implementation should depend on CPU arch. or at least strive to estimate it as much as possible. if, for example, 2 devices are not on same numa node - how can UCP know if it's really a BW limiter or not?

Reply:
@yosefe I discussed this internally with @spotluri and the concern was that if some selection information is lost by not including `ucs_distance_t` in `ucs_sys_dev_distance_t`. One concern that Sreeram is for machines where GPUDirectRDMA shouldn't be used (for instance if there's a GPU on NUMA node 0 but an IB device on only NUMA node 1). By dropping `ucs_distance_t`, we could still yield very low bandwidth and high latency for the pair of (GPU, HCA) but is there a possibility that we still end up using the HCA (and hence GPUDirectRDMA) because of the availability of such a transport. If it's the case that the latency and bandwidth reported by `ucs_sys_dev_distance_t` is ultimately used in protocol selection (which is don't use RNDV put/get but using pipeline path instead), then dropping `ucs_sys_dev_distance_t`  is probably completely safe. Can you confirm if this would be the case? Lmk if I'm not clear.

Reply:
@Akshay-Venkatesh i think we're on the same page. let me clarify:

The plan was that the estimated BW for GPU-direct protocol would be min(port_bw, pci_bw) where pci_bw is the sys/topo-reported number - based on memory and device locality of the specific IB device and current CPU's NUMA node. (for example it could be low if IB device and GPU are not on same NUMA)
On other hand, the estimated BW for pipelined protocol will be min(port_bw, memcpy_bw).

Based on these two numbers, the best protocol would be selected.

Reply:
Sounds good, Yossi. Will drop this field.